src/varnam_state.cpp

```diff
@@ -368,49 +368,79 @@ void VarnamState::selectCandidate(int index) {
+    std::string stringToCommit;
+    std::string wordToLearn;
+    bool isWordBreakKey = isWordBreak(key);
+    bool enablePunctuation = engine_->getConfig()->enablePunctuation.value();
+
+    if (isWordBreakKey) {
+        if (candidates && candidates->cursorIndex() >= 0) {
+            stringToCommit = candidates->candidate(candidates->cursorIndex()).text().toStringForCommit();
+        } else {
+            stringToCommit = preedit_.toStringForCommit();
+        }
+
+        if (enablePunctuation) { // If punctuation processing is enabled
+            buffer_.clear(); // Clear buffer
+            buffer_.push_back(getWordBreakChar(key)[0]); // Add word break char to buffer
+            
+            // Concatenate result word if available
+            if (getVarnamResult() && result_ && varray_length(result_) > 0) {
+                vword *result_word = static_cast<vword *>(varray_get(result_, 0));
+                if (result_word) {
+                    stringToCommit = stringutils::concat(stringToCommit, result_word->text);
+                }
+            }
+        } else {
+            // If punctuation processing is not enabled
+            // Append the punctuation character to stringToCommit
+            stringToCommit += getWordBreakChar(key);
+            // Set wordToLearn to the base candidate text without punctuation
+            wordToLearn = candidates->candidate(candidateSelected).text().toStringForCommit();
+        }
+    } else if (key == FcitxKey_Escape || key == FcitxKey_0 ||
+               !candidates || candidates->size() <= 1 || !result_ || varray_is_empty(result_)) {
+        // Handle escape or special keys
+        stringToCommit.assign(preedit_.toStringForCommit());
+        candidateSelected = 0;
+    } else if ((candidates->cursorIndex() <= 0) && !candidateSelected) {
+        // Handle the case where no candidate is selected
+        vword *first_result = static_cast<vword *>(varray_get(result_, 0));
+        if (first_result) {
+            stringToCommit.assign(first_result->text);
+            candidateSelected = 1;
+        }
+    } else {
+        // Handle regular candidate selection
+        stringToCommit.assign(candidates->candidate(candidateSelected).text().toStringForCommit());
     }
-  } else {
-    stringToCommit.assign(
-        candidates->candidate(candidateSelected).text().toStringForCommit());
-  }
 
-  if (isWordBreak(key)) {
-    stringToCommit = stringutils::concat(stringToCommit, getWordBreakChar(key));
-  }
+      wordToLearn = stringToCommit;
 #ifdef DEBUG_MODE
-  VARNAM_INFO() << "string to commit:" << stringToCommit << " "
-                << getWordBreakChar(key);
+    VARNAM_INFO() << "string to commit:" << stringToCommit << " "
+                  << getWordBreakChar(key);
 #endif
-  ic_->commitString(stringToCommit);
+    ic_->commitString(stringToCommit);
+if (stringToCommit.empty() || lastTypedCharIsDigit || isWordBreakKey ||
```

Comment:
Here we're returning from the function if its word break key right! so the learning part will be skipped. I haven't tested this though, not sure if I missed anything else either. Please check this once.

Reply:
may be i did not tested that. can you check this whole ```commitText``` ? 
````
void VarnamState::commitText(const FcitxKeySym &key) {
    auto candidates = ic_->inputPanel().candidateList();
    std::string stringToCommit;
    std::string wordToLearn;
    bool isWordBreakKey = isWordBreak(key);
    bool enablePunctuation = engine_->getConfig()->enablePunctuation.value();

    if (isWordBreakKey) {
        if (candidates && candidates->cursorIndex() >= 0) {
            stringToCommit = candidates->candidate(candidates->cursorIndex()).text().toStringForCommit();
        } else {
            stringToCommit = preedit_.toStringForCommit();
        }

        if (enablePunctuation) { // If punctuation processing is enabled
            buffer_.clear(); // Clear buffer
            buffer_.push_back(getWordBreakChar(key)[0]); // Add word break char to buffer
            
            // Concatenate result word if available
            if (getVarnamResult() && result_ && varray_length(result_) > 0) {
                vword *result_word = static_cast<vword *>(varray_get(result_, 0));
                if (result_word) {
                    stringToCommit = stringutils::concat(stringToCommit, result_word->text);
                }
            }
        } else {
            // If punctuation processing is not enabled
            // Append the punctuation character to stringToCommit
            stringToCommit += getWordBreakChar(key);
            // Set wordToLearn to the base candidate text without punctuation
            wordToLearn = candidates->candidate(candidateSelected).text().toStringForCommit();
        }
    } else if (key == FcitxKey_Escape || key == FcitxKey_0 ||
               !candidates || candidates->size() <= 1 || !result_ || varray_is_empty(result_)) {
        // Handle escape or special keys
        stringToCommit.assign(preedit_.toStringForCommit());
        candidateSelected = 0;
    } else if ((candidates->cursorIndex() <= 0) && !candidateSelected) {
        // Handle the case where no candidate is selected
        vword *first_result = static_cast<vword *>(varray_get(result_, 0));
        if (first_result) {
            stringToCommit.assign(first_result->text);
            candidateSelected = 1;
        }
    } else {
        // Handle regular candidate selection
        stringToCommit.assign(candidates->candidate(candidateSelected).text().toStringForCommit());
    }

    wordToLearn = stringToCommit;

    ic_->commitString(stringToCommit);

    // Early exit if learning should be skipped
    if (stringToCommit.empty() || 
        ic_->capabilityFlags().test(CapabilityFlag::PasswordOrSensitive) ||
        !engine_->getConfig()->shouldLearnWords.value() || !candidateSelected) {
        reset();
        return;  // Exit early if learning should be skipped
    }

#ifdef DEBUG_MODE
    VARNAM_INFO() << "Original string to commit:" << stringToCommit;  // Log original string
#endif

    // Adjust wordToLearn by removing only the word break character if `isWordBreakKey` is true
    if (isWordBreakKey && enablePunctuation) {
        wordToLearn = wordToLearn.substr(0, wordToLearn.size() - 1);  // Remove the word break character
    }

    // Remove the last character if it's a digit
    if (lastTypedCharIsDigit) {
        wordToLearn = wordToLearn.substr(0, wordToLearn.size() - 1);  // Remove last digit
    }

#ifdef DEBUG_MODE
    VARNAM_INFO() << "Learn word after adjustments:" << wordToLearn;  // Log the adjusted word to learn
#endif

    // Start the learning thread with the word to learn (without word break or digit)
    std::thread learnThread(varnam_learn_word, engine_->getVarnamHandle(),
                            std::move(wordToLearn), 0);
    learnThread.detach();

    reset();  // Reset after learning the word
}
```

Reply:
check new push if that solves this issue